graphql-language/src/parser.rs

```diff
@@ -0,0 +1,363 @@
+#![allow(dead_code, unused_variables)]
+use ast::*;
+use lexer::Punctuator::*;
+use lexer::Token::{
+  Eof,
+  Punctuator,
+  Name as TokenName,
+  IntValue as TokenInt,
+  FloatValue as TokenFloat,
+  StringValue as TokenString,
+};
+use lexer::{Lexer, Token};
+use source::Source;
+use std::iter::Peekable;
+
+pub type Res<'a, T> = Result<T, &'a str>;
+pub type DRes<'a, T> = Res<'a, (T, usize)>;
+
+pub struct Parser<'a> {
+  lexer: Peekable<Lexer<'a>>,
+  source: Source<'a>,
+}
+
+impl<'a> Parser<'a> {
+  pub fn new(src: Source<'a>) -> Self {
```

Comment:
I take this back, but look at the JS [parser object](https://github.com/graphql/graphql-js/blob/master/src/language/parser.js#L1002-L1015) which stores all of the state in the parser. Also, most of these methods could use the signature `parse_*(&mut self)` instead of `parse_*(parser: &mut Parser)`.